File: docs/lucy-targets/author.mjs
#!/usr/bin/env node
// Source authoring only. Rendering uses the existing terrestrial preparation recipe.
import { readFile, writeFile, mkdir, readdir } from 'node:fs/promises';
import { resolve, dirname, relative } from 'node:path';
import { createHash } from 'node:crypto';
import { execFileSync } from 'node:child_process';
import sharp from 'sharp';
import * as fontkit from 'fontkit';
import { createPlanetTitleSource } from '../../tools/prepare-planet-title-sources.mjs';
import { PLANET_TITLE_RECIPE } from '../../src/platform/planet-title-recipe.mjs';
import { parsePdsRadiusTable } from '../../tools/objects/terrestrial-layers/obj-shape.mjs';
import { simplifyRadialShape } from '../../tools/objects/terrestrial-layers/radial-terrain.mjs';
import { renderRadialSnapshot } from '../../tools/objects/terrestrial-layers/radial-snapshot.mjs';
import { paintMissingCoverage } from '../../src/platform/prepare-missing-coverage.mjs';

const root = resolve(import.meta.dirname, '../..');
if (process.cwd() !== root) throw new Error('Run from the repository root.');
const inputsArg = process.argv.find(arg => arg.startsWith('--inputs='));
const inputsPath = inputsArg ? resolve(root, inputsArg.slice(9)) : new URL('./inputs.json', import.meta.url);
const { bodies, checkedOn } = JSON.parse(await readFile(inputsPath));
const commonCommit = 'e97ee9532b17beaf0c7ae38281c5bef12b64fa5b';
// Current page contract loads only metadata into Astro's build graph.
const common = 'src/planets/menoetius/';
const readCommon = path => execFileSync('git', ['show', `${commonCommit}:${common}${path}`]);
const commonJson = path => JSON.parse(readCommon(path));
const hash = bytes => createHash('sha256').update(bytes).digest('hex');
const write = async (path, value) => {
  await mkdir(dirname(path), { recursive: true });
  await writeFile(path, typeof value === 'string' || Buffer.isBuffer(value) ? value : JSON.stringify(value, null, 2) + '\n');
};
const files = async path => (await Promise.all((await readdir(path, { withFileTypes: true })).map(e => e.isDirectory() ? files(resolve(path, e.name)) : [resolve(path, e.name)]))).flat();
const replace = (text, b) => text.replaceAll('menoetius', b.id).replaceAll('Menoetius', b.name);
const configFor = (path, b) => JSON.parse(replace(readCommon(path).toString(), b));
async function repin(b) {
  const pkg = resolve(root, 'src/planets', b.id), src = resolve(pkg, 'source');
  const manifest = JSON.parse(await readFile(resolve(src, 'manifest.json')));
  const declared = new Set([...manifest.inputs, ...manifest.generatedIntermediates].map(e => e.path));
  manifest.documents = [];
  for (const path of (await files(src)).sort()) {
    const rel = relative(src, path);
    if (rel === 'manifest.json' || declared.has(rel)) continue;
    const bytes = await readFile(path);
    manifest.documents.push({ path: rel, expectedBytes: bytes.length, expectedSha256: hash(bytes), purpose: 'Pinned body-specific source interpretation or preparation input.' });
  }
  await write(resolve(src, 'manifest.json'), manifest);
  const descriptorPath = resolve(pkg, 'object.json'), descriptor = JSON.parse(await readFile(descriptorPath));
  for (const ref of descriptor.properties.recipe.sources) ref.sha256 = hash(await readFile(resolve(pkg, ref.path)));
  await write(descriptorPath, descriptor);
}
if (process.argv.includes('--refresh-pins')) {
  for (const b of bodies) await repin(b);
  process.exit(0);
}
const fontPath = resolve(root, 'src/planets/earth/source/presentation/InterVariable.ttf');
if (hash(await readFile(fontPath)) !== PLANET_TITLE_RECIPE.sourceSha256) throw new Error('Shared title font digest differs.');
const font = fontkit.openSync(fontPath).getVariation({ wght: PLANET_TITLE_RECIPE.weight, opsz: PLANET_TITLE_RECIPE.opticalSize });
const neutral = await sharp({ create: { width: 64, height: 32, channels: 3, background: { r: 160, g: 160, b: 160 } } }).png().toBuffer();
const width = 512, height = 256;
const grid = paintMissingCoverage(Buffer.alloc(width * height * 3, 160), { width, height, channels: 3 }, new Uint8Array(width * height).fill(1));
const contextMap = await sharp(grid, { raw: { width, height, channels: 3 } }).ensureAlpha().webp({ lossless: true, effort: 4 }).toBuffer();
for (const b of bodies) {
  const pkg = resolve(root, 'src/planets', b.id), src = resolve(pkg, 'source');
  const [a, c2, c] = b.fullAxesKm.map(x => x / 2), radius = Math.cbrt(a * c2 * c);
  const shape = [];
  for (let lat = -90; lat <= 90; lat += 5) for (let lon = 0; lon <= 360; lon += 5) {
    const p = lat * Math.PI / 180, l = lon * Math.PI / 180;
    const r = 1 / Math.sqrt((Math.cos(p) * Math.cos(l) / a) ** 2 + (Math.cos(p) * Math.sin(l) / c2) ** 2 + (Math.sin(p) / c) ** 2);
    shape.push(`${lon} ${lat} ${r.toFixed(12)}`);
  }
  const table = shape.join('\n') + '\n';
  await write(resolve(src, 'shape/ellipsoid.tab'), table);
  await write(resolve(src, 'material/neutral.png'), neutral);
  const dimensions = b.fullAxesKm.join(' × ') + ' km';
  const coverage = `${b.shapeMeaning} Full approximation dimensions: ${dimensions}. ${b.orientationMeaning} The grid marks unmapped terrain.`;
  let pole = b.poleIcrfDegrees;
  if (!pole) {
    const [lon, lat] = b.poleEclipticDegrees.map(v => v * Math.PI / 180), ob = 23.439291111 * Math.PI / 180;
    const x = Math.cos(lat) * Math.cos(lon), y = Math.cos(lat) * Math.sin(lon), z = Math.sin(lat);
    pole = [(Math.atan2(y * Math.cos(ob) - z * Math.sin(ob), x) * 180 / Math.PI + 360) % 360, Math.asin(y * Math.sin(ob) + z * Math.cos(ob)) * 180 / Math.PI];
  }
  const measurements = {
    schema: 'cssearth-approximate-ellipsoid@1', id: b.id, horizonsCommand: b.horizons,
    source: { url: b.source, papers: b.papers, credit: b.credit, retrieved: checkedOn },
    constraints: { kind: 'approximate-ellipsoid', fullAxesKm: b.fullAxesKm, semiAxesKm: [a, c2, c], fullAxesUncertaintyKm: b.axisUncertaintyKm,
      referenceRadiusKm: radius, axisConvention: 'Full axes are halved exactly. +Z is the adopted pole; +X is the arbitrary display meridian.', qualification: b.shapeMeaning,
      scale: 'The reference radius is the geometric mean of the approximation semiaxes. It is a rendering scale, not an independent measured mean radius or convex-model volume.' },
    approximation: { semiAxesKm: [a, c2, c], assumptions: [b.shapeMeaning, b.orientationMeaning], qualification: 'Observation-constrained ellipsoid; no invented local terrain or albedo.' },
    sampling: { stepDegrees: 5, longitudeDirection: 'east-positive', radiusUnit: 'km', formula: 'r = 1 / sqrt((cos(lat)*cos(lon)/a)^2 + (cos(lat)*sin(lon)/b)^2 + (sin(lat)/c)^2)' },
    spin: { periodHours: b.periodHours, periodText: b.periodText, poleEclipticDegrees: b.poleEclipticDegrees ?? null, poleIcrfRaDecDegrees: pole,
      eclipticToIcrfObliquityDegrees: b.poleEclipticDegrees ? 23.439291111 : null, meridianDegrees: 0, meaning: b.orientationMeaning },
    unresolved: b.unresolved
  };
  await write(resolve(src, 'measurements.json'), measurements);
  await write(resolve(src, 'preparation/rotation.json'), {
    schema: b.periodHours === null ? 'cssearth-display-orientation@1' : 'cssearth-observed-pole@1',
    rightAscensionDegrees: pole[0], declinationDegrees: pole[1], displayMeridianDegrees: 0,
    ...(b.periodHours === null ? {} : { periodHours: b.periodHours }), phase: 'arbitrary-display-phase',
    source: b.id === 'patroclus' ? 'https://doi.org/10.3847/1538-3881/ad1f6e' : b.source,
    coordinateSystem: 'ICRF/J2000; observed or explicitly approximate pole; arbitrary display meridian', qualification: b.orientationMeaning
  });
  const config = configFor('source/preparation/terrestrial.json', b);
  config.geometry.radiusKm = radius;
  config.geometry.camera.framingScale = Math.min(1, radius / a);
  config.geometry.radialTerrain.simplification = { targetFaces: 480, maximumErrorMeters: 25 * radius, method: 'source-meshoptimizer' };
  config.raster.observations[0].metadata.coverage = coverage;
  config.celestial.sunSource = 'JPL Horizons heliocentric state at the fixed 2026-09-03 TT scene epoch. ' + b.orientationMeaning;
  await write(resolve(src, 'preparation/terrestrial.json'), config);
  const content = configFor('source/content/object.json', b);
  content.panel.introduction = b.introduction;
  content.panel.facts = [
    { id: 'shape', label: 'Shape evidence', value: 'Approximate ellipsoid' },
    { id: 'dimensions', label: 'Full approximation dimensions', value: dimensions },
    { id: 'rotation', label: b.id === 'patroclus' ? 'Mutual orbital period' : 'Rotation period', value: b.periodText },
    { id: 'class', label: 'Population', value: b.population ?? 'Jupiter Trojan' }
  ];
  content.panel.moreFacts = [];
  content.lenses.controls[0].description = coverage;
  content.lenses.controls[0].title = b.datasetTitle ?? 'Published shape approximation';
  content.lenses.controls[0].detail = 'Approximate shape';
  content.lenses.controls[0].source.url = b.source;
  content.settings.controls.find(c => c.name === 'shadows').checked = false;
  content.settings.controls.find(c => c.name === 'orbit').checked = false;
  content.resources = [{ label: 'Shape and dimensions', role: 'surface', description: b.credit, href: b.source }, ...b.papers.map(href => ({ label: 'Scientific source', role: 'facts', description: 'Published model and interpretation', href })), content.resources.find(r => r.role === 'stars')];
  content.provenance.editorial = { url: b.source, credit: b.credit };
  content.provenance.physical = { path: '../measurements.json', credit: b.credit };
  await write(resolve(src, 'content/object.json'), content);
  await write(resolve(src, 'presentation/title-mark.json'), { schema: 'cssearth-title-source@1', ...createPlanetTitleSource(b.name, font) });
  await write(resolve(src, 'presentation/minimap.json'), commonJson('source/presentation/minimap.json'));
  for (const file of ['preparation/acquisition.json', 'stars/ESO-IMAGE-LICENSE.md', 'stars/LICENSE.md', 'stars/hyg-v41-field.json']) await write(resolve(src, file), replace(readCommon('source/' + file).toString(), b));
  const mesh = parsePdsRadiusTable(table, config.geometry.radialTerrain.grid);
  const faces = await simplifyRadialShape(mesh, config.geometry.radialTerrain, 230 / (radius * 1000));
  const snapshotRecipe = { generator: 'tools/objects/terrestrial-layers/radial-snapshot.mjs', inputs: ['occultation-approximation', 'model-surface'], size: 512, longitudeDegrees: 55, latitudeDegrees: 20, ambient: .45, diffuse: .55, lensId: 'model' };
  const context = await renderRadialSnapshot({ ...snapshotRecipe, faces, map: contextMap });
  await write(resolve(src, 'presentation/context.png'), context);
  const navigation = configFor('source/preparation/navigation.json', b);
  Object.assign(navigation.source, { origin: b.source, credit: b.credit, expectedBytes: context.length, expectedSha256: hash(context), recipe: snapshotRecipe });
  await write(resolve(src, 'preparation/navigation.json'), navigation);
  const manifest = configFor('source/manifest.json', b);
  for (const entry of manifest.inputs.filter(e => ['occultation-approximation', 'model-surface'].includes(e.id))) {
    const bytes = await readFile(resolve(src, entry.path));
    Object.assign(entry, { expectedBytes: bytes.length, expectedSha256: hash(bytes), origin: b.source, credit: b.credit, coverage, licenseEvidence: [b.source] });
    entry.projection.referenceRadiusMeters = radius * 1000;
  }
  manifest.generatedIntermediates = [{ ...navigation.source }];
  await write(resolve(src, 'manifest.json'), manifest);
  const descriptor = configFor('object.json', b);
  delete descriptor.properties.worldFrame;
  descriptor.properties.recipe.shape.radiusKm = radius;
  descriptor.prepared.sha256 = '0'.repeat(64);
  descriptor.properties.page = { stylesheets: [`src/renderers/css/styles/${b.id}-surfaces.css`] };
  await write(resolve(pkg, 'object.json'), descriptor);
  await repin(b);
  // Keep the body README and its reviewed evidence when regenerating source data.
  await write(resolve(pkg, 'NOTICE.md'), `# Credits\n\nScientific shape constraints: ${b.credit}. Numerical extraction and ellipsoid approximation: cssEarth, MIT. Retain the source citations and approximate status; research papers are not relicensed or bundled. The missing-data grid is authored display content, not observed regolith. ESO/S. Brunier panorama: CC BY 4.0. Inter: SIL OFL 1.1. HYG metadata: Astronexus, CC BY-SA 4.0. See source/stars for full license records.\n`);
  await write(resolve(root, `src/renderers/css/styles/${b.id}-surfaces.css`), replace(execFileSync('git', ['show', `${commonCommit}:src/renderers/css/styles/menoetius-surfaces.css`]).toString(), b));
  await write(resolve(root, `tests/objects/browser/${b.id}/browser-profile.mjs`), replace(execFileSync('git', ['show', `${commonCommit}:tests/objects/browser/menoetius/browser-profile.mjs`]).toString(), b));
  console.log(JSON.stringify({ id: b.id, sourceFaces: mesh.indices.length, outputFaces: faces.length, radiusKm: radius, contextBytes: context.length }));
}

File: tools/objects/source-authoring/distant-worlds/author.py
"""Extract the selected published numeric models; shared preparers own rendering.

Run from the repository root after restoring pinned source inputs. This is the
radial ellipsoid extraction updates existing packages and their current bindings.
"""
from pathlib import Path
import hashlib, json, math, shutil, sys

ROOT = Path(__file__).resolve().parents[4]
INPUT_PATH = Path(sys.argv[1]) if len(sys.argv) > 1 else Path('tools/objects/source-authoring/distant-worlds/inputs.json')
INPUTS = json.loads((ROOT / INPUT_PATH).read_text())
REFERENCE_ROOT = ROOT / INPUTS.get('referenceDirectory', 'output/distant-worlds/references')

def write(path, value):
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(value if isinstance(value, str) else json.dumps(value, ensure_ascii=False, indent=2) + '\n')

def pin(path):
    data = path.read_bytes()
    return dict()

def current(path, body):
    return json.loads((ROOT / 'src/objects' / body['id'] / path).read_text())

# Source identities are authored in the current package before extraction.
reviewed = {}
for body in INPUTS['bodies']:
    path = ROOT / 'src/objects' / body['id'] / 'source/manifest.json'
    if not path.exists():
        raise ValueError(f'Author source identities and bindings in {path} before extracting a new body.')
    manifest = json.loads(path.read_text())
    if manifest.get('schema') != f"css{body['id']}-authoritative-sources@2":
        raise ValueError(f'Update {path} to the current Sources contract before extraction.')
    expected_ids = [entry['id'] for entry in manifest['inputs']]
    expected_ids += ['published-shape', 'model-surface']
    existing_inputs = {entry['id']: entry for entry in manifest['inputs']}
    for local_id in expected_ids:
        if not existing_inputs.get(local_id, {}).get('sourceBinding'):
            raise ValueError(f'Author the source binding for {body["id"]}/{local_id} before extraction.')
    reviewed[body['id']] = manifest

for body in INPUTS['bodies']:
    ident, name, radius = body['id'], body['name'], body['radiusKm']
    package = ROOT / 'src/objects' / ident
    source = package / 'source'
    description = body['shapeMeaning'] + ' ' + body['orientationMeaning'] + ' The grid marks unmapped terrain.'
    config = current('source/preparation/terrestrial.json', body)
    config['geometry']['radiusKm'] = radius
    config['geometry']['camera']['framingScale'] = min(1, 2 * radius / max(body['fullAxesKm']))
    terrain = config['geometry']['radialTerrain']
    terrain['simplification']['maximumErrorMeters'] = radius * 25
    config['raster']['observations'][0]['metadata']['coverage'] = description
    config['celestial']['sunSource'] = 'JPL Horizons fixed 2026-09-03 epoch. ' + body['orientationMeaning']
    manifest = current('source/manifest.json', body)
    manifest['schema'] = f'css{ident}-authoritative-sources@2'
    manifest['documents'] = reviewed[ident]['documents']
    manifest['generatedIntermediates'] = reviewed[ident]['generatedIntermediates']
    manifest['inputs'] = [entry for entry in manifest['inputs'] if entry['id'] not in ['published-shape', 'model-surface']]
    acquisition = current('source/preparation/acquisition.json', body)
    for op in acquisition['operations']: op['groups'] = ['restore', 'refresh']
    axes = [value / 2 for value in body['fullAxesKm']]
    assert abs(math.prod(axes) ** (1/3) - radius) < max(radius * 1e-8, 1e-9)
    lines = []
    for lat in range(-90, 91, 5):
        for lon in range(0, 361, 5):
            p, l = math.radians(lat), math.radians(lon)
            r = 1 / math.sqrt((math.cos(p)*math.cos(l)/axes[0])**2 + (math.cos(p)*math.sin(l)/axes[1])**2 + (math.sin(p)/axes[2])**2)
            lines.append(f'{lon} {lat} {r:.12f}')
    write(source / 'shape/ellipsoid.tab', '\n'.join(lines) + '\n')
    manifest['inputs'].append(dict(id='published-shape', path='shape/ellipsoid.tab', **pin(source/'shape/ellipsoid.tab'), origin=body['source'], credit=body['credit'], license='MIT numeric extraction of published scientific facts', licenseEvidence=[body['source']], consumers=['shape'], coverage=description, projection=dict(type='equirectangular',longitudeDirection='east-positive',referenceRadiusMeters=radius*1000)))
    content = current('source/content/object.json', body)
    binding = next(entry for entry in reviewed[ident]['inputs'] if entry['id'] == 'published-shape')['sourceBinding']
    catalogue_id = binding['references'][0]['catalogueId']
    km = lambda x: f'{x:.2f}' if x < 10 else f'{x:.0f}'
    dimension_text = ' × '.join(f'{x*1000:g}' for x in body['fullAxesKm'])+' m' if radius < 1 else ' × '.join(km(x) for x in body['fullAxesKm'])+' km'
    # Every published panel fact names its source record and the measurement it reads.
    doi = body['source'].split('doi.org/', 1)[1] if 'doi.org/' in body['source'] else None
    fact_source = lambda locator: dict(url=body['source'], label=f'Research publication · DOI {doi}' if doi else body['credit'], checked=INPUTS['checkedOn'],
        path='source/measurements.json', catalogueId=catalogue_id, locator=locator)
    content['panel']['facts'] = [dict(id='dimensions',label='Display model extents',value=dimension_text,source=fact_source('/fullAxesKm/0; /fullAxesKm/1; /fullAxesKm/2'))]
    if body.get('rotationFact'):
        content['panel']['facts'].append(dict(id='rotation',label='Rotation',value=body['periodText'],source=fact_source('/periodText')))
    content['panel']['moreFacts'] = []
    lens = content['lenses']['controls'][0]
    for key in ('description', 'title', 'detail', 'summary'): lens.pop(key, None)
    lens['source'].update(id='published-shape',url=body['source'])
    for control in content['settings']['controls']:
        if control['name'] == 'shadows': control['checked'] = False
    content['resources'] = [dict(label='Shape source',role='surface',description=body['credit'],href=body['source'])] + [dict(label='Scientific source',role='facts',description='Published observations and interpretation',href=url) for url in body['papers']] + [content['resources'][-1]]
    content['provenance']['editorial'] = dict(url=body['source'],credit=body['credit'])
    content['provenance']['physical'] = dict(path='../measurements.json',credit=body['credit'])
    content['displayName'] = body.get('titleLabel', name)
    write(source/'content/object.json',content)
    # Reader text stays outside source/; pnpm prepare:text checks it and flags filler for review.
    citation = dict(catalogueId=catalogue_id, url=body['source'], label=body['credit'], checked=INPUTS['checkedOn'])
    # Extra reader-text citations, such as an orbit fact quoted from JPL, come from the input table.
    text_sources = [*body.get('textCitations', []), citation]
    write(package/'text.json', dict(schema='cssearth-object-text@1', objectId=ident,
        card=dict(text=body.get('card', body['introduction']), sources=text_sources),
        introduction=dict(text=body['introduction'], sources=text_sources),
        datasets={lens['id']: dict(title=body['shapeLabel'], detail='Inferred shape', summary=body.get('datasetSummary', body['shapeMeaning']), sources=[citation])}))
    write(source/'preparation/terrestrial.json',config)
    write(source/'preparation/rotation.json',dict(schema='cssearth-display-orientation@1',rightAscensionDegrees=body['poleIcrfDegrees'][0],declinationDegrees=body['poleIcrfDegrees'][1],displayMeridianDegrees=0,phase='arbitrary-display-phase',source=body.get('poleSource',body['source']),qualification=body['orientationMeaning']))
    write(source/'measurements.json',dict(schema='cssearth-distant-world-model@1',checkedOn=INPUTS['checkedOn'],**body))
    for kind in ['elements','vectors']:
        record = ROOT / 'packages/astronomy/tools/.cache/horizons' / f'asteroid-{kind}-{ident}.txt'
        if record.exists():
            write(source/'reference'/f'horizons-{kind}.txt',record.read_text())
    # Citations do not need downloaded pages. Only explicit file inputs are restored.
    for ref in body.get('references',[]):
        if not ref.get('file'): continue
        target = source / 'reference' / ref['file']
        cached = REFERENCE_ROOT / ref['file']
        if cached.exists():
            target.parent.mkdir(parents=True,exist_ok=True)
            shutil.copyfile(cached,target)
        if not target.exists(): raise FileNotFoundError(target)
        actual=pin(target)
        if not ref.get('retainOriginal'):
            operation = dict(kind='download',groups=['restore','refresh'],path='reference/'+ref['file'],url=ref['url'])
            acquisition['operations'] = [op for op in acquisition['operations'] if op.get('path') != operation['path']] + [operation]
    write(source/'preparation/acquisition.json',acquisition)
    for file in ['material/neutral.png','presentation/InterVariable.ttf']:
        target=source/file
        if not target.exists():
            available=ROOT/'src/objects/annefrank/source'/file
            if not available.exists(): raise FileNotFoundError(f'Restore common input: {available}')
            target.parent.mkdir(parents=True,exist_ok=True)
            shutil.copyfile(available,target)
    manifest['inputs'].append(dict(id='model-surface',path='material/neutral.png',origin='https://github.com/layoutit/cssEarth',credit='cssEarth missing-coverage grid',license='MIT',consumers=['surfaces'],width=64,height=32,lensId='model',label='Shape model',falseColor=False,coverage='Authored neutral material; no observed imagery.',projection=dict(type='equirectangular',longitudeDirection='east-positive',referenceRadiusMeters=radius*1000)))
    for entry in manifest['inputs']:
        entry.setdefault('acquisition',INPUTS.get('acquisitionNote','Restore pinned originals through acquisition; reproduce authored numbers with tools/objects/source-authoring/distant-worlds/author.py.'))
        entry.setdefault('redistribution','Retain source attribution and model qualifications; upstream papers are not relicensed.')
    previous = {entry['id']: entry for entry in reviewed[ident]['inputs']}
    for entry in manifest['inputs']:
        if entry['id'] not in previous or 'sourceBinding' not in previous[entry['id']]:
            raise ValueError(f"Author the source binding for {ident}/{entry['id']} before extraction.")
    manifest['inputs'] = [{**previous[entry['id']], **entry,
        'sourceBinding': previous[entry['id']]['sourceBinding']} for entry in manifest['inputs']]
    write(source/'manifest.json',manifest)
    descriptor=current('object.json',body)
    descriptor['properties']['recipe']['shape']['radiusKm']=radius
    write(package/'object.json',descriptor)
    print(ident,'source-authored')
